### PR TITLE
Open blocks (fence gates, trap doors) with right click

### PR DESCRIPTION
**Pull request by Tonodus**
A new BlockType that supports opening the blocks if the player right clicks them.
It uses the Openable-MaterialData to open/close.

### CODE REVIEW

**Comment by EntityReborn** [comment]
As this is a relatively small update, might be worth it to squash your commits.

**Comment by Tonodus** [comment]
Sorry, do you mean I should squash these four commits into one? I'm going to do that (next time?). I'm relatively new with git, didn't noticed that woul be better.

**Comment by EntityReborn** [comment]
Yes, that is what I mean. Regardless, there isn't an established guideline for this for PRs, besides commits getting squashed right before integration. Up to you.

**Comment by Tonodus** [comment]
Issue: Fence gates don't open themselves off from the player

**Comment by Tonodus** [comment]
Fixed issue, but feel free to comment my solution.

File: src/main/java/net/glowstone/block/blocktype/BlockFenceGate.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import net.glowstone.block.GlowBlock;
+import net.glowstone.entity.GlowPlayer;
+import org.bukkit.block.BlockFace;
+import org.bukkit.block.BlockState;
+import org.bukkit.material.Gate;
+import org.bukkit.material.MaterialData;
+import org.bukkit.util.Vector;
+
+/**
+ * Created by Tonodus (http://tonodus.github.io) on 08.09.2014.
```
Comment (daboross) [inline src/main/java/net/glowstone/block/blocktype/BlockFenceGate.java]: I don't _think_ we want to include things like this. Could you remove this line?

**Comment by daboross** [comment]
Here's a guide to squashing, if you need one: http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html.

File: src/main/java/net/glowstone/block/blocktype/BlockFenceGate.java (hunk 1 above)
Comment (Tonodus) [inline src/main/java/net/glowstone/block/blocktype/BlockFenceGate.java]: I apologize, that's just my IDE's default header I read over.

**Comment by Tonodus** [comment]
Removed header and squashed commits.

**Comment by EntityReborn** [comment]
This currently has merge conflicts. Please resolve.

**Comment by Tonodus** [comment]
Done

**Comment by turt2live** [comment]
Sorry - bot fail :(

**Comment by turt2live** [comment]
@Tonodus this works for standard fence gates and trap doors, but not the extended gates. You still end up placing a block in front of the varied gates (birch, etc)

**Comment by Tonodus** [comment]
@turt2live: Everything should be ok now, sorry for the inconveniences.

**Comment by SpaceManiac** [comment]
Looks pretty good for the most part, just a couple of things with the fence gates:
- The orientation should only be changed when the gate is opened (otherwise there's an animation tic when closing a fence gate from some directions).
- I _think_ the gate shouldn't be reoriented for certain player angles (near perpendicular to the gate's axis) and its old orientation should just be used, at least from what I can see from opening it from this perspective, but I'm not certain.

**Comment by Tonodus** [comment]
@SpaceManiac I fixed all you mentioned:
- I couldn't reproduce your animation tic, but the orientation is now only changed if the gate is actually opened
- You were right, the gate opens itself in the opposite direction the player is facing, but ONLY if the player is facing exactly in this direction. Otherwise the gate is opened to the previous direction. Look at the current code, for more information; should be clear then.

**Comment by SpaceManiac** [comment]
I was trying to change `BlockFenceGate` to use the existing `getOppositeBlockFace` instead of its own `blockFaceFromYaw`, but in doing so noticed this oddity in Bukkit (that was causing this attempt to fail):
https://github.com/GlowstoneMC/Glowkit/blob/master/src/main/java/org/bukkit/material/Gate.java#L29

Any thoughts? Should this be changed Glowkit-side so it isn't so confusing, or left to sit and overridden in BlockFence instead?

**Comment by Tonodus** [comment]
Well, I'd prefer changing the method in Glowkit, so that if we set facingDirection to NORTH it is actually faced to north. The amount of plugins that are broken by this change should be minimal.

**Comment by socram8888** [comment]
@Tonodus: I was also implementing gates when I noticed your PR here. Since this looks prettier than mine, I've only submitted the Glowkit part, which should be useful here: https://github.com/GlowstoneMC/Glowkit/pull/26/files

**Comment by turt2live** [comment]
Bugs from testing this:
- Iron trap doors cannot be placed on the "upper" half of the block. 
  - Don't worry about redstone functionality. But placeholder code for open/close should be there

And one other bug (that is also present in vanilla): 

Opening the fence gate from a perpendicular position then crossing "around" the gate to get to the other side to open on a shallow angle can cause the gate to open the "wrong" way. This is a mathematical error and I don't think it's even worth fixing to be honest. (Graphic included below for a visual representation of what I am talking about)

image

**Comment by Tonodus** [comment]
1a: Added IronTrapDoors
1b: Donno what placeholder code for open/close I should write.. We still can just use the MaterialData Openable to call setOpen()/isOpen().
2: it was harder to implement that bug, than making it "correct". If you do it correct there will be a client side glitch (see comments above). I can change it back to the "right" approach of course, if you prefer that.

**Comment by turt2live** [comment]
As-is, it is the "right" approach as it matches the vanilla server. I just mentioned it so others, if they experience it, don't complain about it. It's a non-issue.

**Comment by SpaceManiac** [comment]
I don't understand what `BlockTrapDoor` is accomplishing here. Why not just have `BlockWoodenTrapDoor` and `BlockIronTrapDoor` _extend_ `BlockTrapDoor` instead of the weirdness you're doing right now?

**Comment by turt2live** [comment]
@Tonodus Your PR is now not mergable due to recent changes. Please update your PR as soon as you can, thanks!

**Comment by dequis** [comment]
Trivial conflict in ItemTable.java, just use both sides.

Merged branch here https://github.com/dequis/Glowstone/compare/pr-200 (which shouldn't even be needed, tbh).

**Comment by Tonodus** [comment]
@SpaceManiac 

> Why not just have BlockWoodenTrapDoor and BlockIronTrapDoor extend BlockTrapDoor

_Short answer:_
Because BlockWoodenTrapDoor extends already BlockOpenable.

_Long answer:_
The current BlockType-system is insufficient. It depends on inheritance instead of composition, what causes the issues as mentioned above: We're running into 'diamond of death's.
Here it's the classes BlockTrapDoor/BlockOpenable, but there are plenty cases where we need composition, as inheritance is insufficient:
F.e. many blocks drop themselves if the block they're attached to is broken. But this is a totally different feature of this block compared to the remain, e.g. block-growth/block-ticking. Both will likely be summarized in superclasses, but the concret class (e.g. BlockSapling) needs both of them.

**Comment by SpaceManiac** [comment]
Merged at long last in 9f19dc9d1ad0b54c589dcb9e6ea09c06a4eae0d0.

**Pull request closed without merging**